src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1314,6 +1329,7 @@ impl Store {
       .await?
       .walk(&mut |path, entry| match entry {
         directory::Entry::File(f) => files.push((path.to_owned(), f.digest(), f.is_executable())),
+        directory::Entry::Symlink(_) => (),
```

Comment:
Sooo, this one is tricky. This is asking for `Files`, which would be symlink oblivious. So I think that you'd need to actually walk the symlink in this case to determine whether the destination was a file.

But also: if the destination is a `Directory`, then you'd need to recurse with a new relative path.

----

I see two options for these cases (where a `DigestTrie` needs to actually "act" like a filesystem):
1. adding another implementation of `DigestTrie::walk` that is symlink oblivious (i.e. will never return a symlink, and instead will walk within the destination of the symlink)
2. Adding methods to `VFS` to implement walking, and/or using the equivalent of `vfs.expand(PathGlobs(["**"]))`, and using them when you want to be symlink oblivious.

The former might be easier? It would be lightly redundant with path globs expansion, but on the other hand, pretending that something is POSIX when it isn't doesn't do you many favors.

Reply:
I added a param to `walk`, which should force callers to think about which behavior they want.

Reply:
Hmmm there's also the issue of "if its symlink oblivious and the symlink points outside of the trie"

Reply:
Seems fine to me to allow dangling symlinks. The target of the symlink is the responsibility of the code invoking the `Process`. The VFS code should just do what it is told to do since it has no semantic knowledge of what the `Digest` is being used for.

Reply:
On the other hand, it could be good to be defensive about this until someone identifies a use-case where they need a dangling symlink. The spec supports them, but leaving them as a TODO that we error for rather than partially implemented seems safer.

Reply:
Welllllll... I actually have a use-case :joy: 

I'm unpacking libs where one has a symlink to a lib in a dependency. So in reality the digests for libA will be merged with its dependencies and the link won't dangle. But in isolation its a dangling link.

In this case, we get a choice If the user tell us they are `oblivious` we could skip dangling symlinks or still return the symlink (so "oblivious" means "follow if possible"). I err on the side of "skipping"

Reply:
Sorry: "dangling but pointing inside of the snapshot root" is one thing, and probably fine.

But "pointing outside of the root of the snapshot", either by being absolute, or doing a relative path lookup outside of the root is what I'm referring to (trying to) ban initially.

Reply:
In this case it'd be relative to the root, but the target is not in the Trie.

Reply:
In general though this code only needs to handle symlinks whose targets are not entries in the Trie.

I'm pretty sure other code handles parsing the target as relative and relative to sandbox root.

Reply:
>But "pointing outside of the root of the snapshot", either by being absolute, or doing a relative path lookup outside of the root is what I'm referring to (trying to) ban initially.

Agreed. We should ban symlink targets escaping the sandbox.

Reply:
> > But "pointing outside of the root of the snapshot", either by being absolute, or doing a relative path lookup outside of the root is what I'm referring to (trying to) ban initially.
> 
> Agreed. We should ban symlink targets escaping the sandbox.

Yep.

> I'm pretty sure other code handles parsing the target as relative and relative to sandbox root.

Yea... I think that as long as the snapshot capture code continues to not support a symlink escaping the root (`VFS::expand` -> `Snapshot::from_path_stats`), we're good.

Reply:
Marking as resolved. Added the param to `walk` for aware/oblivious and made each caller aware (for now, let's visit them inline)